sklearn/datasets/_base.py

```diff
@@ -1454,28 +1464,161 @@ def _fetch_remote(remote, dirname=None, n_retries=3, delay=1):
+
+        checksum = _sha256(temp_file_path)
+        if remote.checksum is not None and remote.checksum != checksum:
+            raise OSError(
+                f"The SHA256 checksum of {remote.filename} ({checksum}) "
+                f"differs from expected ({remote.checksum})."
+            )
+    except BaseException:
+        os.unlink(temp_file.name)
+        raise
+
+    # The following renaming is atomic whenever temp_file_path and
+    # file_path are on the same filesystem. This should be the case most of
+    # the time, but we still use shutil.move instead of os.rename in case
+    # they are not.
+    shutil.move(temp_file_path, file_path)
 
-    file_path = remote.filename if dirname is None else join(dirname, remote.filename)
-    while True:
-        try:
-            urlretrieve(remote.url, file_path)
-            break
-        except (URLError, TimeoutError):
-            if n_retries == 0:
-                # If no more retries are left, re-raise the caught exception.
-                raise
-            warnings.warn(f"Retry downloading from url: {remote.url}")
-            n_retries -= 1
-            time.sleep(delay)
-
-    checksum = _sha256(file_path)
-    if remote.checksum != checksum:
-        raise OSError(
-            "{} has an SHA256 checksum ({}) "
-            "differing from expected ({}), "
-            "file may be corrupted.".format(file_path, checksum, remote.checksum)
-        )
     return file_path
+
+
+def _slugify(value):
+    """Derive a name that is safe to use as filename from the given string.
+
+    Adapted from
+    https://github.com/django/django/blob/master/django/utils/text.py
+
+    Convert to ASCII if 'allow_unicode' is False. Convert spaces or repeated
+    dashes to single dashes. Remove characters that aren't alphanumerics,
+    underscores, or hyphens. Convert to lowercase. Also strip leading and
+    trailing whitespace, dashes, and underscores.
+
+    Note: this version keeps "." characters unchanged contrary to the django
+    version and replace other un-authorized characters by "_".
+    """
+    value = (
+        unicodedata.normalize("NFKD", value).encode("ascii", "ignore").decode("ascii")
+    )
+    value = re.sub(r"[^.\w\s-]", "_", value.lower())
+    value = re.sub(r"_+", "_", value)
+    return re.sub(r"[-\s]+", "-", value).strip("-_")
+
+
+def _derive_folder_and_filename_from_url(url):
+    parsed_url = urlparse(url)
+    path = parsed_url.path
+    if not path:
+        path = "/"
+
+    if "/" in path:
+        base_folder, filename = path.rsplit("/", 1)
+
+    if not filename:
+        filename = "downloaded_file"
+
+    base_folder = _slugify(base_folder)
+    if base_folder:
+        base_folder = "/" + base_folder
+
+    return (
+        _slugify(parsed_url.hostname) + base_folder,
+        _slugify(filename),
+    )
+
+
+def fetch_file(
+    url, folder=None, local_filename=None, sha256=None, n_retries=3, delay=1
+):
+    """Fetch a file from the web.
```

Comment:
Maybe we want to mention that it will be cached on the disk in this short summary.